genie-agent/src/main/java/com/netflix/genie/agent/execution/services/impl/S3ArchivalServiceImpl.java

```diff
@@ -0,0 +1,172 @@
+/*
+ *
+ *  Copyright 2018 Netflix, Inc.
+ *
+ *     Licensed under the Apache License, Version 2.0 (the "License");
+ *     you may not use this file except in compliance with the License.
+ *     You may obtain a copy of the License at
+ *
+ *         http://www.apache.org/licenses/LICENSE-2.0
+ *
+ *     Unless required by applicable law or agreed to in writing, software
+ *     distributed under the License is distributed on an "AS IS" BASIS,
+ *     WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *     See the License for the specific language governing permissions and
+ *     limitations under the License.
+ *
+ */
+
+package com.netflix.genie.agent.execution.services.impl;
+
+import com.amazonaws.services.s3.AmazonS3;
+import com.amazonaws.services.s3.AmazonS3URI;
+import com.amazonaws.services.s3.model.ObjectMetadata;
+import com.amazonaws.services.s3.model.PutObjectRequest;
+import com.netflix.genie.agent.execution.exceptions.ArchivalException;
+import com.netflix.genie.agent.execution.services.ArchivalService;
+import com.netflix.genie.common.internal.jobs.JobConstants;
+import lombok.extern.slf4j.Slf4j;
+import org.springframework.context.annotation.Lazy;
+import org.springframework.stereotype.Component;
+
+import javax.validation.constraints.NotNull;
+import java.io.ByteArrayInputStream;
+import java.io.File;
+import java.net.URI;
+
+/**
+ * Implementation of ArchivalService for S3.
+ *
+ * @author standon
+ * @since 4.0.0
+ */
+@Component
+@Lazy
+@Slf4j
+class S3ArchivalServiceImpl implements ArchivalService {
+
+    private final AmazonS3 amazonS3;
+
+    S3ArchivalServiceImpl(final AmazonS3 amazonS3) {
+        this.amazonS3 = amazonS3;
+    }
+
+    /**
+     * Store the contents of a file to S3.
+     * For a directory all its contents are recursively traversed. Each of the file is
+     * stored as a an object in S3. Each empty directory is also stored as an object with
+     * empty content.
+     * <p>
+     * Example 1
+     * <p>
+     * For a dir hierarchy
+     * /tmp/${jobid}/genie/applications (empty dir)
+     * /tmp/${jobid}/genie/run (file)
+     * /tmp/${jobid}/genie/commands/presto (file)
+     * <p>
+     * dir to be archived - /tmp/${jobid}
+     * target location - s3://$bucketName/{$jobid}
```

Comment:
I think adding `${jobId} in this example is just confusing, since the service itself does nothing with it.
For file `x` uploaded to `/foo/bar`, this will create `/foo/bar/x`.
Bucket is also presented as a variable for some reason, but it's really not, as far as this service is concerned.

If you really want to keep these variables. Be consistent with syntax:
`$foo`, `{$foo}`, `${foo}`

Reply:
I am a bit confused. In the previous chat we had regarding the comment you wanted me to provide an example where the example is more like how a genie job folder actually looks like. Hence {$jobId}. Lets have a quick chat and clarify @mprimi

Reply:
Back then your service was behaving differently so old comment doesn't apply anymore.